src/libzfs_core/py_zfs_core_module.c

```diff
@@ -1,54 +1,686 @@
+	Py_END_ALLOW_THREADS
+
+	Py_DECREF(logmsg);
+	return err ? B_FALSE : B_TRUE;
+}
+
+
+PyDoc_STRVAR(py_zfs_core_exception__doc__,
+"ZFSCoreException(exception)\n"
+"-----------------------\n\n"
+"Python wrapper around error information returned by libzfs core\n\n"
+"attributes:\n"
+"-----------\n"
+"code: int\n"
+"    libzfs errno (one of ZFSError) or regular errno\n"
+"name: str\n"
+"    human-readable name of the libzfs errno\n"
+"errors: tuple | None\n"
+"    tuple of tuples containing error information returned by libzfs_core\n\n"
+"NOTE: the libzfs error may wrap around conventional OS errno. In this case\n"
+"it will be mapped to equivalent libzfs errno, but if that's not possible the\n"
+"libzfs errno will be set to EZFS_UNKNOWN and strerror output written to\n"
+"the error description field.\n"
+);
+static
+void setup_zfs_core_exception(PyObject *module)
+{
+	PyObject *dict = NULL;
+	PyObject *errno_mod = NULL;
+	pylibzfs_core_state_t *state = get_lzc_mod_state(module);
+
+	dict = Py_BuildValue("{s:i,s:s,s:s,s:O}",
+			     "code", EZFS_UNKNOWN,
+			     "msg", "",
+			     "name", "",
+			     "errors", Py_None);
+	PYZFS_ASSERT(dict, "Failed to set up ZFSCoreException");
+
+	state->zc_exc = PyErr_NewExceptionWithDoc(PYLIBZFS_MODULE_NAME
+						  "lzc.ZFSCoreException",
+						  py_zfs_core_exception__doc__,
+						  PyExc_RuntimeError,
+						  dict);
+
+	Py_DECREF(dict);
+	PYZFS_ASSERT(state->zc_exc, "Failed to set up ZFSCoreException");
+
+	/* Get reference to errno.errorcode to convert errno to name */
+	errno_mod = PyImport_ImportModule("errno");
+	PYZFS_ASSERT(errno_mod, "Failed to import errno module.");
+
+	state->errorcode = PyObject_GetAttrString(errno_mod, "errorcode");
+	PYZFS_ASSERT(state->errorcode, "Failed to retrieve errorcode dict.");
+	Py_DECREF(errno_mod);
+}
+
+static
+void set_zfscore_exc(PyObject *module,
+		     const char *msg,
+		     int code,
+		     PyObject *errors_tuple)
+{
+	pylibzfs_core_state_t *state = get_lzc_mod_state(module);
+	PyObject *v = NULL;
+	PyObject *attrs = NULL;
+	const char *name = NULL;
+	PyObject *error_name = NULL;
+	PyObject *pycode = NULL;
+	int err;
+
+	pycode = Py_BuildValue("i", code);
+	if (pycode == NULL)
+		return;
+
+	name = zfs_error_name(code);
+	if (strcmp(name, "UNKNOWN") == 0) {
+		error_name = PyDict_GetItem(state->errorcode, pycode);
+		if ((error_name == NULL) && PyErr_Occurred()) {
+			Py_DECREF(pycode);
+			return;
+		} else if (error_name != NULL) {
+			Py_INCREF(error_name);
+		}
+	}
+
+	/*
+	 * NULL is possible in following scenarios
+	 * 1. errorcode lacks an entry for the errno
+	 * 2. ZFS errno was encoded in nvlist
+	 */
+	if (error_name == NULL) {
+		error_name = PyUnicode_FromString(name);
+		if (error_name == NULL) {
+			Py_DECREF(pycode);
+			return;
+		}
+	}
+
+	v = PyObject_CallFunction(state->zc_exc, "s", msg);
+	if (v == NULL)
+		return;
+
+	/* This steals references to pycode and error_name */
+	attrs = Py_BuildValue(
+		"(NsNO)",
+		pycode,
```

Comment:
Do we need to do `Py_DECREF(pycode)` below in error conditions and in the end before returning (in case of success). We are doing it above in erroneous conditions but not below.